pkg/container/ring_reader.go

```diff
@@ -64,20 +66,28 @@ func (r *RingReader) Next() *v1.Event {
 // position by one. If there are no more event to read, NextFollow blocks
 // until the next event is added to the ring or the context is cancelled.
 func (r *RingReader) NextFollow(ctx context.Context) *v1.Event {
-	if r.c == nil {
+	// if the context changed between invocations, we also have to restart
+	// readFrom, as the old readFrom instance will be using the old context.
+	if r.c == nil || r.ctx != ctx {
 		r.c = r.ring.readFrom(ctx, r.idx)
+		r.ctx = ctx
 	}
-	// when the ring is not full, nil events may be sent to the channel
-	// one should keep reading in such a case
 	for {
```

Comment:
If the `ring.readFrom()` does not return `nil` events anymore (it blocks until it can read a non-nil event), then this for loop his not needed anymore.

Reply:
Good catch! Fixed!